test/v2/services/RetrieveSelfAssessmentChargeHistoryServiceSpec.scala

```diff
@@ -0,0 +1,113 @@
+    RetrieveSelfAssessmentChargeHistoryRequest(
+      nino = nino,
+      transactionId = "anId"
+    )
+
+  val chargeHistoryDetails: ChargeHistoryDetail =
+    ChargeHistoryDetail(
+      taxYear = Some("2019-20"),
+      transactionId = "X123456790A",
+      transactionDate = "2019-06-01",
+      description = "Balancing Charge Debit",
+      totalAmount = 600.01,
+      changeDate = "2019-06-05",
+      changeReason = "Example reason"
+    )
+
+  val retrieveChargeHistoryResponse: RetrieveSelfAssessmentChargeHistoryResponse =
+    RetrieveSelfAssessmentChargeHistoryResponse(
+      chargeHistoryDetails = Seq(chargeHistoryDetails)
+    )
+
+  trait Test extends MockRetrieveSelfAssessmentChargeHistoryConnector {
+
+    implicit val hc: HeaderCarrier              = HeaderCarrier()
+    implicit val logContext: EndpointLogContext = EndpointLogContext("RetrieveSelfAssessmentChargeHistoryController", "RetrieveSelfAssessmentChargeHistory")
+
+    val service = new RetrieveSelfAssessmentChargeHistoryService(
+      connector = mockRetrieveSelfAssessmentChargeHistoryConnector
+    )
+
+  }
+
+  "RetrieveSelfAssessmentChargeHistoryService" when {
+    "service call successful" must {}
+    "return mapped result" in new Test {
+
+      val connectorResponse: RetrieveSelfAssessmentChargeHistoryResponse = retrieveChargeHistoryResponse
+
+      MockRetrieveSelfAssessmentChargeHistoryConnector
+        .retrieveSelfAssessmentChargeHistory(requestData)
+        .returns(Future.successful(Right(ResponseWrapper(correlationId, connectorResponse))))
+
+      await(service.retrieveChargeHistory(requestData)) shouldBe Right(ResponseWrapper(correlationId, connectorResponse))
+    }
+  }
+
+  "unsuccessful" must {
+    "map errors according to spec" when {
+
+      def serviceError(desErrorCode: String, error: MtdError): Unit =
+        s"a $desErrorCode error is returned from the service" in new Test {
+
+          MockRetrieveSelfAssessmentChargeHistoryConnector
+            .retrieveSelfAssessmentChargeHistory(requestData)
+            .returns(Future.successful(Left(ResponseWrapper(correlationId, DownstreamErrors.single(DownstreamErrorCode(desErrorCode))))))
+
+          await(service.retrieveChargeHistory(requestData)) shouldBe Left(ErrorWrapper(correlationId, error))
+        }
+
+      val input: Seq[(String, MtdError)] = Seq(
+        ("INVALID_CORRELATIONID", InternalError),
+        ("INVALID_IDTYPE", InternalError),
+        ("INVALID_IDVALUE", NinoFormatError),
+        ("INVALID_REGIME_TYPE", InternalError),
+        ("INVALID_DOC_NUMBER", TransactionIdFormatError),
+        ("INVALID_DATE_FROM", InternalError),
+        ("INVALID_DATE_TO", InternalError),
+        ("INVALID_DATE_RANGE", InternalError),
+        ("REQUEST_NOT_PROCESSED", InternalError),
+        ("NO_DATA_FOUND", NotFoundError),
+        ("SERVER_ERROR", InternalError),
+        ("SERVICE_UNAVAILABLE", InternalError)
+      )
```

Comment:
Also need to map INVALID_REQUEST to InternalError (seems to be missing from our spec).

Reply:
I have added this error type now. Thank you.